### PR TITLE
Optimize _collect_new_io_core in extract_model

### PR DESCRIPTION
**Pull request by tonypottera24**
### Description

Optimized the `_collect_new_io_core` function.
The updated version behaves the same way as the previous one.

The modification includes:
1. Use a for loop to build `new_io_tensors`
2. Add return type
3. Moved the duplicate check to `extract_model` function.

### Motivation and Context

The original version uses `set`, `map` and `dict` to make simple works complicated.

### CODE REVIEW

**Review by justinchuby — APPROVED** [review]

**Comment by justinchuby** [comment]
Thanks!

File: onnx/utils.py [hunk 1]
```diff
@@ -23,26 +23,23 @@ def __init__(self, model: ModelProto) -> None:
     def _build_name2obj_dict(objs):  # type: ignore
         return {obj.name: obj for obj in objs}
 
-    def _collect_new_io_core(self, original_io, io_names_to_extract):  # type: ignore
+    def _collect_new_io_core(
+        self, original_io, io_names_to_extract
```
Comment (justinchuby) [inline onnx/utils.py:27]: Could you help add types to the inputs as well?

File: onnx/utils.py [hunk 2]
```diff
@@ -23,26 +23,23 @@ def __init__(self, model: ModelProto) -> None:
...
-
-        new_io_tensors = [original_io_map[name] for name in io_names_to_keep]
-        # activation become input or output
-        new_io_tensors.extend(self.vimap[name] for name in new_io_names_to_add)
-
-        # adjust sequence
-        new_io_tensors_map = self._build_name2obj_dict(new_io_tensors)
-        return [new_io_tensors_map[name] for name in io_names_to_extract]
+        new_io_tensors = []
+        for io_name_to_extract in io_names_to_extract:  # assume no duplicates
```
Comment (justinchuby) [inline onnx/utils.py:31]:
```suggestion
        for io_name_to_extract in io_names_to_extract:
```

**Comment by justinchuby** [comment]
I quickly looked and it seems fine, if the test coverage is good? I haven’t check that yet though

File: onnx/utils.py (hunk 1 above)
Comment (tonypottera24) [reply onnx/utils.py]: Thank you @justinchuby , I add the input types
```
original_io: RepeatedCompositeFieldContainer[ValueInfoProto],
io_names_to_extract: list[str],
```

**Comment by tonypottera24** [comment]
> I quickly looked and it seems fine, if the test coverage is good? I haven’t check that yet though

In fact... I think the `output_name` in https://github.com/onnx/onnx/blob/main/onnx/utils.py#L96 should be `input_name`... 
I can try to improve the test in my next PR.

**Pull request merged**